scripts/globals/wsquest.lua

```diff
@@ -0,0 +1,634 @@
+local WSQUEST_CONT2 = 2; -- Player has turned in completed trial weapon and received Map to Annals ('cont2')
+local WSQUEST_FINISH = 3; -- Player has killed NM and received Annals of Truth ('finish')
+
+local function getWsQuestState(quest, player)
+    if (player:getQuestStatus(quest.logId, quest.questId) == QUEST_ACCEPTED) then
+        if (player:hasKeyItem(dsp.ki.ANNALS_OF_TRUTH)) then
+            return WSQUEST_FINISH;
+        elseif (player:hasKeyItem(dsp.ki.MAP_TO_THE_ANNALS_OF_TRUTH)) then
+            return WSQUEST_CONT2;
+        else
+            return WSQUEST_CONT1;
+        end;
+    else
+        local canEquip = player:canEquipItem(quest.trialWeaponId, true);
+        local sufficientSkill = player:getSkillLevel(quest.skillId) >= quest.minSkill;
+        local hasWeapon = player:hasItem(quest.trialWeaponId);
+        local hasTrainingGuide = player:hasKeyItem(dsp.ki.WEAPON_TRAINING_GUIDE);
+
+        if (canEquip and sufficientSkill and not hasWeapon and not hasTrainingGuide) then
+            return WSQUEST_START;
+        end;
+    end;
+
+    return nil;
+end;
+
+local function getWsnmVariableName(quest)
+    return string.format('Killed_%s', string.gsub(quest.wsnmName, ' ', '_'));
+end;
+
+function getWsQuestTradeEvent(quest, player, trade)
+    local wsPoints = (trade:getItem(0):getWeaponskillPoints());
+
+    if (getWsQuestState(quest, player) == WSQUEST_CONT1 and trade:hasItemQty(quest.trialWeaponId, 1) and trade:getItemCount() == 1) then
+        if (wsPoints < 300) then
+            return quest.eventIds.tradedUnfinishedWeapon;
+        else
+            return quest.eventIds.tradedFinishedWeapon;
+        end;
+    end;
+
+    return nil;
+end;
+
+function getWsQuestTriggerEvent(quest, player)
+    local state = getWsQuestState(quest, player);
+
+    if (state == WSQUEST_START) then
+        return quest.eventIds.start; -- WS Quest start
+    elseif (state == WSQUEST_CONT1) then
+        return quest.eventIds.cont1;
+    elseif (state == WSQUEST_CONT2) then
+        return quest.eventIds.cont2;
+    elseif (state == WSQUEST_FINISH) then
+        return quest.eventIds.finish; -- WS Quest completed (WSNM killed)
+    end;
+
+    return nil;
+end;
+
+function handleQmTrigger(quest, player)
+    if (getWsQuestState(quest, player) == WSQUEST_CONT2) then
+        if (player:getVar(getWsnmVariableName(quest)) == 1) then
+            player:messageSpecial(KEYITEM_OBTAINED, dsp.ki.ANNALS_OF_TRUTH);
+            player:addKeyItem(dsp.ki.ANNALS_OF_TRUTH);
+        elseif (GetMobAction(quest.wsnmId) == 0) then
+            player:messageSpecial(SENSE_OMINOUS_PRESENCE);
+            SpawnMob(quest.wsnmId):updateClaim(player);
+        end;
+    else
+        player:messageSpecial(NOTHING_OUT_OF_ORDINARY);
+    end;
+end;
+
+function handleWsnmDeath(quest, player)
+    if (getWsQuestState(quest, player) == WSQUEST_CONT2) then
+        player:setVar(getWsnmVariableName(quest), 1);
```

Comment:
If we're keeping this system as is, I think this would be a perfect use case for local variable, instead of a character variable.

It makes a very temporary (destroyed on zone), and makes it more retail accurate, because if you zone without picking up the key item, you have to kill NM again.

Reply:
local to which block?

Reply:
The lua entity binding setLocalVar: https://github.com/DarkstarProject/darkstar/blob/master/src/map/lua/lua_baseentity.cpp#L541

The variable is only temporarily stored with the entity, and not stored in the database. It would be destroyed when the player zones.